utils: refactor CMake invocation construction

compnerd opened the pull request:
Make the flag handling more explicit for the various platforms. This prepares us to add additional SDKs in the future which require separate handling of flags.

Steelskin commented on utils/build.ps1:
```diff
@@ -1291,202 +1301,267 @@ function Build-CMakeProject {
+            }
+          }
+
+          Add-KeyValueIfNew $Defines CMAKE_C_COMPILER $CC
```
Do we actually ever use values that are not previously set here? I did a quick pass and it seems like this never happens - the only defines that are something not overridden in this function are passed with the same value as would be set here (`CMAKE_SYSTEM_NAME` and `CMAKE_Swift_COMPILER_TARGET`).
I am wondering if we could cache and/or have static maps with the values set for a given configuration rather than re-computing them every time.

compnerd committed:
utils: refactor CMake invocation construction

Make the flag handling more explicit for the various platforms. This
prepares us to add additional SDKs in the future which require separate
handling of flags.

Co-authored-by: Fabrice de Gans <steelskin@gmail.com>

compnerd commented:
@swift-ci please smoke test

compnerd commented:
@swift-ci please smoke test
